relay-threading/benches/pool.rs

```diff
@@ -0,0 +1,159 @@
+    let runtime = create_runtime();
+    let mut group = c.benchmark_group("pool_scaling");
+    group.sampling_mode(criterion::SamplingMode::Flat);
+    group.measurement_time(Duration::from_secs(10));
+
+    // Test with different numbers of threads
+    for threads in [1, 2, 4, 8].iter() {
+        let pool = AsyncPoolBuilder::new(runtime.handle().clone())
+            .num_threads(*threads)
+            .max_concurrency(100)
+            .build()
+            .unwrap();
+
+        // Test with different task counts
+        for tasks in [100, 1000, 10000].iter() {
+            group.bench_with_input(
+                BenchmarkId::new(format!("threads_{}", threads), tasks),
+                tasks,
+                |b, &tasks| {
+                    b.to_async(&runtime).iter(|| run_benchmark(&pool, tasks));
+                },
+            );
+        }
+    }
+
+    group.finish();
+}
+
+fn bench_multi_threaded_spawn(c: &mut Criterion) {
+    let runtime = create_runtime();
+    let mut group = c.benchmark_group("multi_threaded_spawn");
+    group.sampling_mode(criterion::SamplingMode::Flat);
+    group.measurement_time(Duration::from_secs(10));
+
+    // Test with different numbers of spawning threads
+    for spawn_threads in [2, 4, 8].iter() {
+        // Test with different task counts
+        for tasks in [1000, 10000].iter() {
+            group.bench_with_input(
+                BenchmarkId::new(format!("spawn_threads_{}", spawn_threads), tasks),
+                tasks,
+                |b, &tasks| {
+                    b.to_async(&runtime).iter(|| async {
+                        let pool = Arc::new(
+                            AsyncPoolBuilder::new(runtime.handle().clone())
+                                .num_threads(4) // Fixed number of worker threads
+                                .max_concurrency(100)
+                                .build()
+                                .unwrap(),
+                        );
+
+                        let tasks_per_thread = tasks / spawn_threads;
+                        let mut handles = Vec::new();
+
+                        // Spawn tasks from multiple threads
+                        for _ in 0..*spawn_threads {
+                            let runtime = runtime.handle().clone();
+                            let pool = pool.clone();
+                            let handle = std::thread::spawn(move || {
+                                runtime.block_on(run_benchmark(&pool, tasks_per_thread));
+                            });
+                            handles.push(handle);
+                        }
+
+                        // Wait for all spawning threads to complete
+                        for handle in handles {
+                            handle.join().unwrap();
+                        }
+                    });
+                },
+            );
+        }
+    }
+
+    group.finish();
+}
+
+criterion_group!(benches, bench_pool_scaling, bench_multi_threaded_spawn);
```

Comment:
What are these benchmarks benchmarking?

It looks like they would all be dominated by tokio runtime or other external dependencies.

Reply:
Yes, the benchmarks were meant to be twofold:
1. Validate the execution performance mostly to make sure we have no regressions if we were to change it down the line.
2. Implement code to drive the pool with high load to make sure it's functionally correct (this is why we use the atomic count to validate correctness).

If you feel like it's not a proper benchmark I can either remove it or adjust it.